mafia.py

```diff
@@ -1,110 +1,171 @@
-    msg = ""
-    if request.args.get("Kill") is not None:
-        ip = request.args.get("Kill")
-        if ip in ip2player.keys():
-            if ip2player[ip].get_state() == "alive":
-                ip2player[ip].set_state("dead")
-            else:
-                ip2player[ip].set_state("alive")
-        else:
-            return render_template("404.html", is_farsi=True)
-    if request.args.get("Ban") is not None:
-        ip = request.args.get("Ban")
-        if ip in ip2player.keys():
-            if ip2player[ip].get_state() == "alive":
-                ip2player[ip].set_state("banned")
-            elif ip2player[ip].get_state() == "banned":
-                ip2player[ip].set_state("alive")
+def god_page():
+    def kill(player : Player) -> None:
+        # notice banned players cannot be killed
+        if player.is_alive():
+            player.die()
+
+    def switch_ban(player : Player) -> None:
+        player.switch_ban()
+
+    def comment(player : Player) -> None:
+        if player.get_comment() is False:
+            if player_manager.n_comments > max_comments(nPlayers):
+                nonlocal msg
+                msg = "Error: Out of Comments."
+                return
+            player_manager.comment(player)
         else:
-            return render_template("404.html", is_farsi=True)
-    if request.args.get("Comment") is not None:
-        ip = request.args.get("Comment")
-        if ip in ip2player.keys():
-            if ip2player[ip].get_comment() is False:
-                if nComments <= nPlayers // 3:
-                    ip2player[ip].set_comment(True)
-                    nComments += 1
-                    comments_ordered.append(ip)
-                else:
-                    msg = "Error: Out of Comments."
-            else:
-                ip2player[ip].set_comment(False)
-                nComments -= 1
-                comments_ordered.remove(ip)
-        else:
-            return render_template("404.html", is_farsi=True)
-    return render_template("GOD.html", ip2player=ip2player,
-                           prompt_message=msg, roles={role:roles.count(role) for role in set(roles)},
-                           comments=comments_ordered, role2team=role2team)
+            player_manager.uncomment(player)
+
+    request_mapper : Dict[str, Callable] = {
+        "Kill": kill,
+        "Ban": switch_ban,
+        "Comment": comment
+    }
+
+    # filters out invalid requests
+    def get_requests() -> List[Tuple[str, Player]]:
+        requests = []
+        for key in request_mapper.keys():
+            ip = request.args.get(key)
+            
+            if ip is None:
+                continue
+            if not player_manager.is_ip_valid(ip):
+                print("invalid ip / request: ", ip, f"({key})")
+                return -1
+
+            player = player_manager.get_player(ip)
+            requests.append((key, player))
+        return requests
```

Comment:
I don't get the point of this part please explain it for me and it's use in application

Reply:
it's supposed to check for every possible request type in our request object, and check if the ip address is valid (it was previously done by duplicated lines of ip checking).

what you may find strange at first is that this function returns a list of request types and their arguments.
it isn't uncommon to split a big function to smaller ones with different responsibilities, which I did for the god function.
tho the main reason i made get_requests was that fact i was unsure if it is possible for a request object to have multiple args or not (and i actually guess the answer is no, which means i'll probably clean this part a little bit soon)